File: algorithms/medium/merge_nodes_in_between_zeros.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
//
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }

impl Solution {
    pub fn merge_nodes(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut ans = Box::new(ListNode::new(0));
        let mut tail = &mut ans;
        let mut p = head.unwrap().next;
        let mut cur = 0;
        while let Some(node) = p {
            p = node.next;
            if node.val != 0 {
                cur += node.val;
            } else {
                tail.next = Some(Box::new(ListNode::new(cur)));
                if let Some(ref mut next) = tail.next {
                    tail = next;
                }
                cur = 0;
            }
        }
        ans.next
    }
}

File: algorithms/medium/merge_nodes_in_between_zeros.ex
# Definition for singly-linked list.
#
# defmodule ListNode do
#   @type t :: %__MODULE__{
#           val: integer,
#           next: ListNode.t() | nil
#         }
#   defstruct val: 0, next: nil
# end

defmodule Solution do
  @spec merge_nodes(head :: ListNode.t() | nil) :: ListNode.t() | nil
  def merge_nodes(head) do
    aux(nil, 0, head.next)
  end

  def aux(ans, _cur_val, nil), do: reverse(ans)

  def aux(ans, cur_val, %ListNode{val: val, next: next}) do
    if val == 0 do
      aux(%ListNode{val: cur_val, next: ans}, 0, next)
    else
      aux(ans, cur_val + val, next)
    end
  end

  def reverse(lst), do: reverse(lst, nil)

  def reverse(nil, lst), do: lst

  def reverse(%ListNode{val: val, next: next}, lst) do
    reverse(next, %ListNode{val: val, next: lst})
  end
end
